dpctl/memory/_sycl_usm_array_interface_utils.pxi

```diff
@@ -0,0 +1,200 @@
+        return _queue_ref_copy_from_SyclQueue(ptr, <SyclQueue> syclobj)
+    elif type(syclobj) is SyclContext:
+        ctx = <SyclContext>syclobj
+        return _queue_ref_copy_from_USMRef_and_SyclContext(ptr, ctx)
+    elif type(syclobj) is str:
+        q = SyclQueue(syclobj)
+        return _queue_ref_copy_from_SyclQueue(ptr, <SyclQueue> q)
+    elif pycapsule.PyCapsule_IsValid(syclobj, "SyclQueueRef"):
+        q = SyclQueue(syclobj)
+        return _queue_ref_copy_from_SyclQueue(ptr, <SyclQueue> q)
+    elif pycapsule.PyCapsule_IsValid(syclobj, "SyclContextRef"):
+        ctx = <SyclContext>SyclContext(syclobj)
+        return _queue_ref_copy_from_USMRef_and_SyclContext(ptr, ctx)
+    elif hasattr(syclobj, '_get_capsule'):
+        cap = syclobj._get_capsule()
+        if pycapsule.PyCapsule_IsValid(cap, "SyclQueueRef"):
+            q = SyclQueue(cap)
+            return _queue_ref_copy_from_SyclQueue(ptr, <SyclQueue> q)
+        elif pycapsule.PyCapsule_IsValid(cap, "SyclContexRef"):
+            ctx = <SyclContext>SyclContext(cap)
+            return _queue_ref_copy_from_USMRef_and_SyclContext(ptr, ctx)
+        else:
+            return NULL
+    else:
+        return NULL
+
+
+cdef object _pointers_from_shape_and_stride(
+    int nd, object ary_shape, Py_ssize_t itemsize, Py_ssize_t ary_offset,
+    object ary_strides):
+    """
+    Internal utility: for given array data about shape/layout/element
+    compute left-most displacement when enumerating all elements of the array
+    and the number of bytes of memory between the left-most and right-most
+    displacements.
+
+    Returns: tuple(min_disp, nbytes)
+    """
+    if (nd > 0):
+        if (ary_strides is None):
+            nelems = 1
+            for si in ary_shape:
+                sh_i = int(si)
+                if (sh_i <= 0):
+                    raise ValueError("Array shape elements need to be positive")
+                nelems = nelems * sh_i
+            return (ary_offset, nelems * itemsize)
+        else:
+            min_disp = ary_offset
+            max_disp = ary_offset
+            for i in range(nd):
+                str_i = int(ary_strides[i])
+                sh_i = int(ary_shape[i])
+                if (str_i > 0):
+                    max_disp += str_i * (sh_i - 1)
+                else:
+                    min_disp += str_i * (sh_i - 1);
+            return (min_disp, (max_disp - min_disp + 1) * itemsize)
+    elif (nd == 0):
+        return (ary_offset, itemsize)
+    else:
+        raise ValueError("Array dimensions can not be negative")
+
+
+cdef class _USMBufferData:
+    """
+    Internal data struct populated from parsing
+    `__sycl_usm_array_interface__` dictionary
+    """
+    cdef DPCTLSyclUSMRef p
+    cdef int writeable
+    cdef object dt
+    cdef Py_ssize_t itemsize
+    cdef Py_ssize_t nbytes
+    cdef SyclQueue queue
+
+    @staticmethod
+    cdef _USMBufferData from_sycl_usm_ary_iface(dict ary_iface):
```

Comment:
can we expose this to Python iface? I see a reuse in Numba.

Reply:
Yes, we can. Accessor functions for native members of the struct will need to be added though.

Should be in a separate PR though